clang/utils/TableGen/ClangDiagnosticsEmitter.cpp

```diff
@@ -1562,6 +1563,60 @@ void clang::EmitClangDiagsCompatIDs(const llvm::RecordKeeper &Records,
     OS << "DIAG_COMPAT_IDS_END()\n";
 }
 
+/// ClangDiagsIntefaceEmitter - Emit the diagnostics interface header for
+/// a Clang component.
+void clang::EmitClangDiagsInterface(llvm::raw_ostream &OS,
+                                    const std::string &Component) {
+  if (Component.empty())
+    PrintFatalError("'-gen-clang-diags-iface' requires a component name");
+
+  std::string ComponentUpper = StringRef(Component).upper();
+  const char *Comp = Component.c_str();
+  const char *Upper = ComponentUpper.c_str();
+
+  OS << llvm::format(R"c++(
+namespace clang {
```

Comment:
These are all so common (differing only by the one name change) that I wonder if tablegen is the right place for this, instead of just a predefined file.  BUT I can't htink of a way to do that which doesn't require preprocessor string-replacement inside of a `#include` or `define` identifier, which, IIRC is UB?

SO I guess this is ok and the only way we can reasonably do this?

Reply:
> BUT I can't htink of a way to do that which doesn't require preprocessor string-replacement inside of a `#include` or `define` identifier, which, IIRC is UB?

Yeah, I think the preprocessor directives are the main problem, e.g. the `#define ASTSTART`; there isn’t really a way to do this just in the preprocessor unfortunately...